junction/src/main/java/org/killbill/billing/junction/plumbing/billing/DefaultInternalBillingApi.java

```diff
@@ -178,27 +179,63 @@ private void addBillingEventsForBundles(final List<SubscriptionBaseBundle> bundl
                 }
             } else { // billing is not off
                 final SubscriptionBase baseSubscription = subscriptions != null && !subscriptions.isEmpty() ? subscriptions.get(0) : null;
-                addBillingEventsForSubscription(account, subscriptions, baseSubscription, dryRunMode, context, result, skipSubscriptionsSet, catalog);
+                addBillingEventsForSubscription(account, subscriptions, baseSubscription, currentAccountBCD, context, result, skipSubscriptionsSet, catalog);
+            }
+        }
+
+        // If dryRun is specified, we don't want to to update the account BCD value, so we initialize the flag updatedAccountBCD to true
+        if (currentAccountBCD == 0 && !dryRunMode) {
+            BillingEvent oldestAccountAlignedBillingEvent = null;
+
+            for (final BillingEvent event : result) {
+                if (event.getBillingAlignment() != BillingAlignment.ACCOUNT) {
+                    continue;
+                }
+
+                final BigDecimal recurringPrice = event.getRecurringPrice(event.getEffectiveDate());
+                final boolean hasRecurringPrice = recurringPrice != null; // Note: could be zero (BCD would still be set, by convention)
+                final boolean hasUsage = event.getUsages() != null && !event.getUsages().isEmpty();
+                if (!hasRecurringPrice &&
+                    !hasUsage) {
+                    // Nothing to bill, ignored for the purpose of BCD calculation
+                    continue;
+                }
+
+                if (oldestAccountAlignedBillingEvent == null ||
+                    event.getEffectiveDate().compareTo(oldestAccountAlignedBillingEvent.getEffectiveDate()) < 0 ||
+                    (event.getEffectiveDate().compareTo(oldestAccountAlignedBillingEvent.getEffectiveDate()) == 0 && event.getTotalOrdering().compareTo(oldestAccountAlignedBillingEvent.getTotalOrdering()) < 0)) {
+                    oldestAccountAlignedBillingEvent = event;
+                }
+            }
+
+            if (oldestAccountAlignedBillingEvent == null) {
+                return;
+            }
+
+            // BCD in the account timezone
+            final int accountBCDCandidate = oldestAccountAlignedBillingEvent.getBillCycleDayLocal();
+            if (accountBCDCandidate != 0) {
+                log.info("Setting account BCD='{}', accountId='{}'", accountBCDCandidate, account.getId());
+                accountApi.updateBCD(account.getExternalKey(), accountBCDCandidate, context);
+
+                // Because we now have computed the real BCD, we need to re-compute the BillingEvents BCD for ACCOUNT alignments (see BillCycleDayCalculator#calculateBcdForAlignment).
+                // The code could maybe be optimized (no need to re-run the full function?), but since it's run once per account, it's probably not worth it.
```

Comment:
I think this is at least worth a discussion (not strong feeling though): Less about the perf, but more about the maintenance of the code -- this is related to the comment below in  `calculateBcdForTransition`. One thought, is that we could detect first if `currentAccountBCD` is 0 and extract its value from `ACCOUNT` aligned subscriptions, and then proceed with populating billing events without all this complexity.

Reply:
I agree, and this is what I tried to do originally, but it's harder in practice: we need all billing events, retrieve catalog objects and run the `BillCycleDayCalculator` logic to compute that BCD (we effectively need 2 pass, no matter how you look at it).

That being said, I understand the concern. I'll try to do another pass to see if I'm more inspired today 🤔